### PR TITLE
[Feature] RedisSentinel support of RedisLockStore

### PR DESCRIPTION
**Pull request by spennymain**
Fixes: #3924

**Proposed changes**:
- Add Redis Sentinel functionality to the Lock Store

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

# TODO: haven't added any tests -- would like some guidance/recommendations and feedback from the contributors

This is the configuration my rasa deployment is currently using to integrate with my Redis Sentinel Cluster which I have as a service in kubernetes - the `sentinel_url` is currently DNS of the namespace and service which GKE resolves to an IP address so we can deploy the LockStore without a fixed IP

I welcome suggestions but feel that the decomposition pattern is a clean way to integrate Sentinel into the code.

### CODE REVIEW

**Comment by sara-tagger** [comment]
Thanks for submitting a pull request 🚀 @erohmensing will take a look at it as soon as possible ✨

**Comment by wochinge** [comment]
hi @spennymain , thanks for your contribution! I'll give it a review this week 🤞🏻

File: rasa/core/lock_store.py [hunk 1]
```diff
@@ -261,6 +272,39 @@ def save_lock(self, lock: TicketLock) -> None:
         self.red.set(self.key_prefix + lock.conversation_id, lock.dumps())
 
 
+class RedisSentinel():
```
Comment (wochinge) [inline rasa/core/lock_store.py:275]:
```suggestion
class RedisSentinel:
```

File: rasa/core/lock_store.py [hunk 2]
```diff
@@ -261,6 +272,39 @@ def save_lock(self, lock: TicketLock) -> None:
...
+    """Sentinel support for RedisLockStore."""
+
+    def __init__(
```
Comment (wochinge) [inline rasa/core/lock_store.py:278]: Can you please add type annotations and docstrings for this class?

File: rasa/core/lock_store.py [hunk 3]
```diff
@@ -261,6 +272,39 @@ def save_lock(self, lock: TicketLock) -> None:
...
+        self,
+        sentinel_url,
+        sentinel_port,
+        sentinel_name,
+        socket_timeout
+    ):
+        from redis.sentinel import Sentinel
```
Comment (wochinge) [inline rasa/core/lock_store.py:285]: This should in my opinion be a top level import.

File: rasa/core/lock_store.py [hunk 4]
```diff
@@ -261,6 +272,39 @@ def save_lock(self, lock: TicketLock) -> None:
...
+        self._log_connection(sentinel_url, sentinel_port, sentinel_name)
```
Comment (wochinge) [inline rasa/core/lock_store.py:286]: optional: I don't see the benefit of having a separate method for this

File: rasa/core/lock_store.py [hunk 5]
```diff
@@ -261,6 +272,39 @@ def save_lock(self, lock: TicketLock) -> None:
...
+        self.sentinel = Sentinel([(sentinel_url, sentinel_port)], socket_timeout=socket_timeout)
+        self.sentinel_name = sentinel_name
+        self.socket_timeout = socket_timeout
+
+    def find_master(self):
+        return self.sentinel.master_for(self.sentinel_name, socket_timeout=self.socket_timeout)
+
+    def get(self, conversation_id: Text):
+        return self.find_master().get(conversation_id)
```
Comment (wochinge) [inline rasa/core/lock_store.py:295]: that should in my opinion use `find_slave` as it's a read operation.

File: rasa/core/lock_store.py [hunk 6]
```diff
@@ -221,14 +224,22 @@ def __init__(
...
+                host=host,
+                port=int(port),
+                db=int(db),
+                password=password,
+                ssl=use_ssl,
+                socket_timeout=socket_timeout,
+            )
+        else:
+            self.red = RedisSentinel(
+                sentinel_url,
```
Comment (wochinge) [inline rasa/core/lock_store.py:238]:
```suggestion
                sentinel_name,
```

File: rasa/core/lock_store.py [hunk 7]
```diff
@@ -203,6 +203,9 @@ def __init__(
         use_ssl: bool = False,
         key_prefix: Optional[Text] = None,
         socket_timeout: float = DEFAULT_SOCKET_TIMEOUT_IN_SECONDS,
+        sentinel_name: Optional[Text] = "mymaster",
+        sentinel_url: Optional[Text] = None,
+        sentinel_port: Optional[Text] = 26379,
```
Comment (wochinge) [inline rasa/core/lock_store.py:208]:
```suggestion
        sentinel_port: int = 26379,
```

Comment (wochinge) [inline rasa/core/lock_store.py:208]: can you please document the additional parameters in the docstring?

**Review by wochinge — COMMENTED** [review]
Thanks for this awesome contribution. Can you please also add a changelog and update the documentation [here](https://github.com/RasaHQ/rasa/blob/f7942d73fff5e5aab078c0e1654d9242f5245478/docs/docs/lock-stores.mdx#L45)?

> I welcome suggestions but feel that the decomposition pattern is a clean way to integrate Sentinel into the code.

I agree that the pattern is a great choice 👍🏻 

Unfortunately we don't have any integration tests for Redis at the moment. Did your manual tests work as expected?

**Comment by wochinge** [comment]
@spennymain How is it going? Would be great to wrap this up

**Comment by spennymain** [comment]
Hi @wochinge! 

I dont have any tests other than running in a k8s instance with bitnami redis-sentinel helm chart

I can set aside sometime to update the docs and changelog this week. 

What do you think ?

**Comment by wochinge** [comment]
> I dont have any tests other than running in a k8s instance with bitnami redis-sentinel helm chart

Ok. I think we should at least test that the class correctly tries to use the sentinel if the sentinel config is provided by the user.

**Comment by wochinge** [comment]
@spennymain Are you still planning to wrap this up? I close this for now - please feel free to re-open if you want to continue working on it.

**Pull request closed without merging**